[DoctrineBridge] Reject fields UniqueEntity cannot query with the default repository method

wkania opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

`findBy()` reads any array criteria value as an `IN (...)` list rather than as a single value routed through the field's type, so a `UniqueEntity` constraint on an array-valued field built a query that could never match. Captured SQL on the base branch:

| field | value | generated SQL |
| --- | --- | --- |
| `simple_array` | `['foo','bar']` | `WHERE simpleArray IN (?, ?)` with both parameters NULL |
| `json` | `['key' => 'value', 'list' => [1,2,3]]` | `WHERE json IN (?, ?, ?, ?)`, the nested array flattened |
| `json` | `'plain'` | `WHERE json = '"plain"'`, correct |

The failure is silent: no rows match, so no violation is reported and the application believes it has a uniqueness rule that it does not.

To-many associations fail differently. `$em->initializeObject()` throws `Class "Doctrine\Common\Collections\ArrayCollection" is not a valid entity or mapped super class` for a plain `ArrayCollection`, which is the normal case for an unpersisted entity being validated in a form, and `findBy()` cannot query the inverse side of an association at all.

Both are now rejected with a `ConstraintDefinitionException` pointing at the `repositoryMethod` option, which is the supported way to check such fields. To-many fields also stop being passed to `initializeObject()`, so `repositoryMethod` finally works for them.

Reworked during review, with two changes of approach:

* **the check is on the runtime value, not the declared column type.** Keying off `Types::SIMPLE_ARRAY` and `Types::JSON` missed DBAL's own `JSON_OBJECT`, `JSONB` and `JSONB_OBJECT`, and every user-defined type whose PHP representation is an array. It also rejected `json` columns holding a scalar, which work correctly today. Checking `is_array($fieldValue)` covers all of them and leaves the working case alone;
* **no DBAL dependency.** `doctrine/dbal` is `require-dev` only for this bridge; the runtime requirement is `doctrine/persistence`, and `UniqueEntityValidator` is deliberately persistence-agnostic. A class-constant fetch autoloads, so referencing `Types` made every `UniqueEntity` validation fatal on a MongoDB ODM install with no DBAL. The association check likewise uses `ClassMetadata::isCollectionValuedAssociation()` from `doctrine/persistence` rather than the ORM's `ClassMetadata` constants, so it also applies to ODM.

The result is 14 added lines and no new imports.

Tests: an array-valued field and a to-many association both raise the new exception and both fail on a clean 8.2, the second with the old Doctrine mapping error. A `json` column holding a scalar is asserted to keep validating, which is the case the earlier type-based check regressed. Two small fixtures were needed for the to-many case, since the bridge had no `OneToMany` entity; both pass `SchemaValidator`.

wkania committed:
[DoctrineBridge] Fix UniqueEntityValidator for fields (simple_array, json, ONE_TO_MANY ,MANY_TO_MANY)

stof commented on src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php:
```diff
@@ -1429,4 +1433,187 @@ public function testUuidIdentifierWithSameValueDifferentInstanceDoesNotCauseViol
+        $this->assertCount(1, $this->context->getViolations());
+    }
+
+    #[IgnoreDeprecations]
```
Why making this a legacy test ? Is this testing a deprecated behavior ? Would it be a test that we delete in 8.0 ?
Adding new legacy tests is quite suspicious for a PR that does not introduce any new deprecations.

wkania replied:
- All tests have DoctrineStyle variation.
- Since [PR](https://github.com/symfony/symfony/pull/54744) DoctrineStyle is deprecated, but will be removed in Symfony 8.
- Since [PR](https://github.com/symfony/symfony/pull/58370) DoctrineStyle tests are marked as #[IgnoreDeprecations]

wkania committed:
[DoctrineBridge] Reject fields UniqueEntity cannot query with the default repository method

findBy() reads any array criteria value as an IN list rather than as a single
value routed through the field type, so a UniqueEntity constraint on a
simple_array or json field built a query that could never match and silently
reported no violation. A to-many association fared worse: initializeObject()
threw a Doctrine mapping error on a plain ArrayCollection, and findBy() cannot
query the inverse side at all.

Both are now rejected with a message pointing at the repositoryMethod option,
which is the supported way to check such fields.

The test is on the runtime value rather than on the declared column type, so it
also covers jsonb, user defined array types, and any driver: the bridge only
requires doctrine/persistence, and reading DBAL type constants here would make
the validator fatal on an ODM-only install.
